tests/unit/test_networks.py

```diff
@@ -0,0 +1,209 @@
+# © Crown Copyright GCHQ
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Tests for network implementations.
+
+The tests within this file verify that the network classes and functionality throughout
+the codebase produce the expected results on simple examples.
+"""
+
+import unittest
+
+import optax
+from jax import random
+
+import coreax.networks
+import coreax.util
+
+
+class TestTrainState(unittest.TestCase):
+    """
+    Tests related to :func:`create_train_state`.
+
+    .. note::
+        Learning rate is not tested as it is part of weight updates, not initialisation
+        of the code written in this codebase.
+    """
+
+    def setUp(self):
+        """
+        Generate data for use across unit tests.
+        """
+        self.hidden_dimension = 2
+        self.data_dimension = 5
+        self.state_key = random.key(1989)
+        self.learning_rate = 0.1
+        self.optimiser = optax.adamw
+
+    def test_create_train_state_negative_hidden_dimension(self):
+        """
+        Test create_train_state with a negative hidden dimension size.
+        """
+        score_network = coreax.networks.ScoreNetwork(
+            -self.hidden_dimension, self.data_dimension
+        )
+
+        # Create a train state with this network - we expect optax to catch the invalid
+        # hidden dimension argument and raise an appropriate error
+        with self.assertRaises(TypeError) as error_raised:
```

Comment:
Use `assertRaisesRegex` to combine this and line 69. Similarly for the other tests.

No need to include the full error message to test against, just enough to be sure we've hit the right one. No idea how stable the error wordings are between package versions. If you do want to match the full message, enclose the regex in `^...$` to match the start and end of the expressions.

Reply:
Agreed and updated - I now just check that it discusses non-negative indices

Reply:
Updated other tests to reflect this